go-controller/pkg/ovn/obj_retry.go

```diff
@@ -99,104 +100,140 @@ func (r *retryObjs) addRetryObj(obj interface{}) {
+	entry := r.ensureRetryEntryLocked(key, &retryObjEntry{Mutex: sync.Mutex{}, oldObj: obj, config: config,
+		timeStamp: time.Now(), backoffSec: 1, ignore: true})
+	entry.timeStamp = time.Now()
+	entry.oldObj = obj
+	if entry.config == nil {
+		entry.config = config
+	}
+	if noRetryAdd {
+		// will not be retried for addition
+		entry.newObj = nil
 	}
+	entry.Unlock()
 }
 
 // addDeleteToRetryObj adds an old object that needs to be cleaned up to a retry object
 // includes the config object as well in case the namespace is removed and the object is orphaned from
 // the namespace
 func (r *retryObjs) addDeleteToRetryObj(obj interface{}, key string, config interface{}) {
-	r.retryMutex.Lock()
-	defer r.retryMutex.Unlock()
-	if entry, ok := r.entries[key]; ok {
+	entry := r.ensureRetryEntryLocked(key, nil)
+	if entry != nil {
 		entry.oldObj = obj
 		entry.config = config
+		entry.Unlock()
 	}
 }
 
 // removeDeleteFromRetryObj removes any old object from a retry entry
 func (r *retryObjs) removeDeleteFromRetryObj(key string) {
-	r.retryMutex.Lock()
-	defer r.retryMutex.Unlock()
-	if entry, ok := r.entries[key]; ok {
+	entry := r.ensureRetryEntryLocked(key, nil)
+	if entry != nil {
 		entry.oldObj = nil
 		entry.config = nil
+		entry.Unlock()
 	}
 }
 
 // unSkipRetryObj ensures an obj is no longer ignored for retry loop
 func (r *retryObjs) unSkipRetryObj(key string) {
-	r.retryMutex.Lock()
-	defer r.retryMutex.Unlock()
-	if entry, ok := r.entries[key]; ok {
+	entry := r.ensureRetryEntryLocked(key, nil)
+	if entry != nil {
 		entry.ignore = false
+		entry.Unlock()
 	}
 }
 
 // deleteRetryObj deletes a specific entry from the map
 func (r *retryObjs) deleteRetryObj(key string, withLock bool) {
 	if withLock {
-		r.retryMutex.Lock()
-		defer r.retryMutex.Unlock()
+		entry := r.ensureRetryEntryLocked(key, nil)
+		if entry != nil {
+			defer entry.Unlock()
+		}
 	}
+	r.retryMutex.Lock()
 	delete(r.entries, key)
+	r.retryMutex.Unlock()
 }
 
 // skipRetryObj sets a specific entry from the map to be ignored for subsequent retries
 func (r *retryObjs) skipRetryObj(key string) {
-	r.retryMutex.Lock()
-	defer r.retryMutex.Unlock()
-	if entry, ok := r.entries[key]; ok {
+	entry := r.ensureRetryEntryLocked(key, nil)
+	if entry != nil {
 		entry.ignore = true
+		entry.Unlock()
 	}
 }
 
 // checkRetryObj returns true if an entry with the given key exists, returns false otherwise.
 func (r *retryObjs) checkRetryObj(key string) bool {
+	entry := r.ensureRetryEntryLocked(key, nil)
+	if entry != nil {
+		entry.Unlock()
+		return true
+	}
+	return false
+}
+
+// returns the retryEntry structure with lock held on it. all users of retryEntry should call
+// this function. the mutex and the `ignore` field ensures that the pod add/update/delete do not
+// step on each other. if newRetryEntry is specified, then the function will try to create a
+// new retryEntry in the map, otherwise it will return nil.
```

Comment:
NIT: I would make the comment even more clear:
_if newRetryEntry is specified and the map has no entry with the input key, then the function will try to add newRetryEntry to the map_

Reply:
accept